README.md
# Combine PVOutput Systems
Simple Python3 script to combine two PVOutput Systems continously (normally once per 5 minutes) into a target PVOutput System.

PVOutput also includes possibilities for combining systems, but that needs a donation. With a donation you well get also other functionality, so make a choice if the donation features of PVOutput is a better route.

Notes
* you need 3 PVOutput accounts, for each system one.
* only between 5 and 23 hour data is fetched and combined into the target PVOutput
* the script will exit outside 5 and 23
* if one system is not updated anymore, the target PVOutput system will NOT be updated or the time must be after 22:00 hours (assumption tyhat one system has gone to sleep already)
* the script relies on Python version 3.7 and higher, using the ordered dictionary functionality.

## PVOutput
[PVOutput](https://pvoutput.org/) is a free online service for sharing and comparing photovoltaic solar panel output data. It provides both manual and automatic data uploading facilities.

Output data can be graphed, analysed and compared with other pvoutput contributors over various time periods. The ability to compare with similar systems within close proximity allows both short and longer term performance issues to be easily identified. While PVOutput is primarily focused on monitoring energy generation, it also provides equally capabable facilities to upload and monitor energy consumption data from various energy monitoring devices.

The python script requires multiple PVOutput API_KEY and SYSTEM_ID to function.
* Login in PVOutput and goto your Settings page for each account
* Select Enabled for API Access
* Click on New Key to generate your API key
* Make a note of your System Id
* Save your settings

## Configuration
Change in combine_two_pvoutput_systems.cfg the following lines with your above obtained secrets:
* pvoutput_source1_api_key = a930b6cfe4c326b0b15b83bc3501ddf123456789
* pvoutput_source1_system_id = 12345
* pvoutput_source2_api_key = 0f2dd8190d00369ec893b059034dde1123456789
* pvoutput_source2_system_id = 12346
* pvoutput_target_api_key = 6d7be6f183fc02c53f5792d1ec37180123456789
* pvoutput_target_system_id = 12347

## Usage
### Windows 10
python combine_two_pvoutput_systems.py

### Raspberry pi
combine_two_pvoutput_systems.py script runs on my Raspberry pi Model B with 512 MB memory (yes, very old, from 2012) with Raspbian GNU/Linux 11 (bullseye).

### Raspberry pi Configuration
Steps:
* create a directory solis in your home directory
* copy pvoutput.sh, combine_two_pvoutput_systems.py and combine_two_pvoutput_systems.cfg in this solis directory
* change inside combine_two_pvoutput_systems.cfg the API_KEY and SYSTEM_ID secrets
* chmod + x pvoutput.sh
* add the following line in your crontab -e:

```
4 5 * * * ~/solis/pvoutput.sh > /dev/null
@reboot sleep 234 && ~/pvoutput/pvoutput.sh > /dev/null
```

### log files
Log files are written in the home subdirectory solis
* pvoutput.log containing the processing and data send to PVOutput (and maybe error messages)
* pvoutput.crontab.log containing the crontab output (normally it will say that solis.sh is running).

### Example output pvoutput.log

```
20220805 08:15:06: Target   read_in_memory since 05:00: written=05:00, written wh=0, wh=152, line=20220805,08:05,152,294,,,,410.0
20220805 08:15:06: Last Written target: 08:05
20220805 08:15:06: Source 1 read_in_memory since 08:05: written=08:05, written wh=107, wh=124, line=20220805,08:10,124,204,,,,234.4
20220805 08:15:07: Source 2 read_in_memory since 08:05: written=08:05, written wh=45, wh=53, line=20220805,08:10,53,100,,,,175.6
20220805 08:15:07: Sleeping for 5 minutes
20220805 08:20:08: Source 1 read_in_memory since 08:05: written=08:05, written wh=107, wh=141, line=20220805,08:15,141,204,,,,234.6
20220805 08:20:08: Source 2 read_in_memory since 08:05: written=08:05, written wh=45, wh=62, line=20220805,08:15,62,110,,,,179.6
20220805 08:20:08: Target datapoints: 2
1: data=20220805,08:10,177,304,,,,410
2: 20220805,08:15,203,314,,,,414
20220805 08:25:09: Source 1 read_in_memory since 08:15: written=08:15, written wh=141, wh=161, line=20220805,08:20,161,240,,,,234.8
20220805 08:25:10: Source 2 read_in_memory since 08:15: written=08:15, written wh=62, wh=71, line=20220805,08:20,71,110,,,,179.5
20220805 08:25:10: Target datapoints: 1
1: data=20220805,08:20,232,350,,,,414
20220805 08:30:11: Source 1 read_in_memory since 08:20: written=08:20, written wh=161, wh=178, line=20220805,08:25,178,204,,,,234.3
20220805 08:30:12: Source 2 read_in_memory since 08:20: written=08:20, written wh=71, wh=100, line=20220805,08:25,100,130,,,,179.6
20220805 08:30:12: Target datapoints: 1
1: data=20220805,08:25,278,334,,,,414
20220805 08:35:13: Source 1 read_in_memory since 08:25: written=08:25, written wh=178, wh=196, line=20220805,08:30,196,216,,,,234.0
20220805 08:35:14: Source 2 read_in_memory since 08:25: written=08:25, written wh=100, wh=107, line=20220805,08:30,107,90,,,,167.6
20220805 08:35:14: Target datapoints: 1
1: data=20220805,08:30,303,306,,,,402
```



CHANGELOG.md
<a name="R1.4.0"></a>
# [Fix socket.timeout issue #8 (R1.4.0)](https://github.com/ZuinigeRijder/CombinePVOutputSystems/releases/tag/R1.4.0) - 18 Sep 2022

Fix for issue https://github.com/ZuinigeRijder/SolisCloud2PVOutput/issues/8
- catch socket.timeout and any other exception
- increase timeout from 10 to 30 (seconds)


[Changes][R1.4.0]


<a name="R1.3.0"></a>
# [Fixed problem when PVOutput returns NaN for voltage (R1.3.0)](https://github.com/ZuinigeRijder/CombinePVOutputSystems/releases/tag/R1.3.0) - 28 Aug 2022

Solved issue https://github.com/ZuinigeRijder/CombinePVOutputSystems/issues/3

- When voltage is NaN take previous voltage
- Wait till next 5 minutes interval plus 3 minutes to have possibly faster results

[Changes][R1.3.0]


<a name="R1.2.0"></a>
# [Increased sending retries, endless retry getting data, small improvements (R1.2.0)](https://github.com/ZuinigeRijder/CombinePVOutputSystems/releases/tag/R1.2.0) - 13 Aug 2022

- increased sending retries to 30 minutes
- endless retry getting data
- improved function naming and return value typing

[Changes][R1.2.0]


<a name="R1.1.0"></a>
# [API Secrets configuration moved out of python script into .cfg file (R1.1.0)](https://github.com/ZuinigeRijder/CombinePVOutputSystems/releases/tag/R1.1.0) - 05 Aug 2022

Change your API SECRETS in the configuration file "combine_two_pvoutput_systems.cfg". This way a new python script will not overwrite your API SECRETS. 

[Changes][R1.1.0]


<a name="R1.0.0"></a>
# [First release (R1.0.0)](https://github.com/ZuinigeRijder/CombinePVOutputSystems/releases/tag/R1.0.0) - 05 Aug 2022

- less logging
- added TODAY constant as yyyymmdd to compute it only once
- made pylint compliant
- make flake8 and black linters compliant
- fixed endless loop when no results in read in memory

[Changes][R1.0.0]


[R1.4.0]: https://github.com/ZuinigeRijder/CombinePVOutputSystems/compare/R1.3.0...R1.4.0
[R1.3.0]: https://github.com/ZuinigeRijder/CombinePVOutputSystems/compare/R1.2.0...R1.3.0
[R1.2.0]: https://github.com/ZuinigeRijder/CombinePVOutputSystems/compare/R1.1.0...R1.2.0
[R1.1.0]: https://github.com/ZuinigeRijder/CombinePVOutputSystems/compare/R1.0.0...R1.1.0
[R1.0.0]: https://github.com/ZuinigeRijder/CombinePVOutputSystems/tree/R1.0.0

<!-- Generated by https://github.com/rhysd/changelog-from-release v3.7.0 -->


CombinePVOutputSystems.py
# == CombinePVOutputSystems.py Author: Zuinige Rijder ========================
import time
import sys
from datetime import datetime
from urllib.error import HTTPError, URLError
from urllib.request import urlopen, Request

# == Secrets, fill in yours ==================================================
PVOUTPUT_SOURCE1_API_KEY = 'xxxx'
PVOUTPUT_SOURCE1_SYSTEM_ID = 'xxxx'

PVOUTPUT_SOURCE2_API_KEY = 'xxxx'
PVOUTPUT_SOURCE2_SYSTEM_ID = 'xxxx'

PVOUTPUT_TARGET_API_KEY = 'xxxx'
PVOUTPUT_TARGET_SYSTEM_ID = 'xxxx'

# == Constants ===============================================================
PVOUTPUT_GET_URL = 'https://pvoutput.org/service/r2/getstatus.jsp'
PVOUTPUT_ADD_URL = 'http://pvoutput.org/service/r2/addbatchstatus.jsp'

GET_SOURCE1_URL = PVOUTPUT_GET_URL + \
    '?key=' + PVOUTPUT_SOURCE1_API_KEY + \
    '&sid=' + PVOUTPUT_SOURCE1_SYSTEM_ID + \
    '&h=1&asc=0&limit=288'
GET_SOURCE2_URL = PVOUTPUT_GET_URL + \
    '?key=' + PVOUTPUT_SOURCE2_API_KEY + \
    '&sid=' + PVOUTPUT_SOURCE2_SYSTEM_ID + \
    '&h=1&asc=0&limit=288'
GET_TARGET_URL = PVOUTPUT_GET_URL + \
    '?key=' + PVOUTPUT_TARGET_API_KEY + \
    '&sid=' + PVOUTPUT_TARGET_SYSTEM_ID + \
    '&h=1&asc=0&limit=288'
TARGET_HEADER = {
    'X-Pvoutput-Apikey': PVOUTPUT_TARGET_API_KEY,
    'X-Pvoutput-SystemId': PVOUTPUT_TARGET_SYSTEM_ID,
    'Content-Type': 'application/x-www-form-urlencoded',
    'Accept': 'text/plain'
}


# == log =====================================================================
def log(msg):
    print(datetime.now().strftime("%Y%m%d %H:%M:%S") + ': ' + msg)


# == printSplitted ===========================================================
def printSplitted(content):
    count = 0
    for item in content.split(';'):
        count = count + 1
        print(str(count) + ': ' + item)


# ==  getOrPost===============================================================
def getOrPost(prefix, url, request) -> str:
    log(prefix)
    errorstring = ''
    try:
        with urlopen(request, timeout=10) as response:
            body = response.read()
            content = body.decode("utf-8")
            # log('Content=')
            # printSplitted(content)
            return content
    except HTTPError as error:
        errorstring = str(error.status) + ': ' + error.reason
    except URLError as error:
        errorstring = str(error.reason)
    except TimeoutError:
        errorstring = 'Request timed out'

    log('ERROR: ' + url + ' -> ' + errorstring)
    time.sleep(60)  # retry after 1 minute

    return 'ERROR'


# == get =====================================================================
def get(prefix, url) -> str:
    request = Request(url)
    content = getOrPost('GET ' + prefix, url, request)
    return content


# == post ====================================================================
def post(url, data, header) -> str:
    post_data = data.encode("utf-8")
    request = Request(url, data=post_data, headers=header)
    content = getOrPost('POST Target', url, request)
    return content


# == addbatchstatus ==========================================================
def addBatchStatus(dataPointsCount, dataPointsStr) -> str:
    log('Target datapoints: ' + str(dataPointsCount))
    printSplitted(dataPointsStr)
    # return 'OK'
    retry = 0
    while True:
        retry = retry + 1
        content = post(PVOUTPUT_ADD_URL, dataPointsStr, TARGET_HEADER)
        if content != 'ERROR' or retry > 10:
            return content


# == getStatus ===============================================================
def getStatus(prefix, url, dateString, timeString) -> str:
    parameters = '&d="' + dateString + '"&t="' + timeString + '"'
    requestUrl = url + parameters
    content = get(prefix + ' ' + parameters, requestUrl)
    return content


# == computeMinutes ==========================================================
def computeMinutes(dateString, content) -> int:
    minutes = 300  # start at 5 hours
    if len(content) > 14 and content[0:8] == dateString:
        minutes = round(int(content[9:11]) * 60 + int(content[12:14]))
    # print('computeMinutes(' + content[9:14] + ') = ' + str(minutes))
    return minutes


# == minutesToTimeString =====================================================
def minutesToTimeString(minutes) -> str:
    hours = int(minutes / 60)
    min = round(minutes - (60 * hours))
    timeStr = "%02d" % hours + ':' + "%02d" % min
    # print('minutesToTimeString(' + str(minutes) + ') = ' + timeStr)
    return timeStr


# == readInMemory ============================================================
def readInMemory(prefix, content, dateString, targetMinutes, lastWrittenWh):
    sinceTime = minutesToTimeString(targetMinutes)
    log(prefix + ' readInMemory since: ' + sinceTime)
    minutes2result = {}
    splitted = content.split(';')
    prevLine = ''
    prevWh = lastWrittenWh
    lastWrittenTime = sinceTime
    i = len(splitted) - 1
    while i >= 0:
        line = splitted[i]
        # print('Lineno ', i, ': ', line)
        i -= 1

        d, t, whStr, energyEfficiency, w, \
            averagePower, normalisedOutput, energyConsumption, \
            PowerConsumption, temp, v = \
            line.split(',')
        if d != dateString:  # skip dates not today
            continue  # not interested in other days

        wh = int(whStr)
        if v == 'NaN':
            v = '0'

        minutes = computeMinutes(dateString, line)
        if minutes % 5 != 0:
            sys.exit('Minutes not in 5 minutes increments: ' + line)

        line = d + ',' + t + ',' + str(wh) + ',' + w + ',,,,' + v
        if minutes > targetMinutes:
            if prevLine == '':
                prevMinutes = minutes - 5  # no missing entries
            else:
                prevMinutes = computeMinutes(dateString, prevLine)
                minutes2result[prevMinutes] = prevLine

            if prevMinutes + 5 != minutes:
                print('Missing entry:\n---> ' + prevLine + '\n---> ' + line)
                missingEntries = int(((minutes - prevMinutes - 5) / 5))
                # print('Number of missing entries: ' + str(missingEntries))
                wh = int(prevWh + ((wh - prevWh) / (1 + missingEntries)))
                t = minutesToTimeString(prevMinutes + 5)
                addedLine = d + ',' + t + ',' + str(wh) + ',' + w + ',,,,' + v

                print('Prev : ' + prevLine)
                print('Added: ' + addedLine)
                print('Next : ' + line + '\n')
                minutes2result[prevMinutes + 5] = addedLine
                line = addedLine
                i += 1  # do this entry again next loop
            else:
                minutes2result[minutes] = line
                # print('Entry: ' + line)
        else:
            lastWrittenTime = minutesToTimeString(minutes)
            lastWrittenWh = wh

        prevLine = line
        prevWh = wh

    log(prefix +
        ' readInMemory last values: written time=' + lastWrittenTime +
        ', written wh=' + str(lastWrittenWh) + ', wh=' + str(prevWh) +
        ', line=' + prevLine)
    return minutes2result, lastWrittenWh


# == MAIN ====================================================================

# get the latest combined target time
now = datetime.now()
dateStr = now.strftime("%Y%m%d")
target = getStatus('Target  ', GET_TARGET_URL, dateStr, '05:00')
lastTMinutes = computeMinutes(dateStr, target)
source1Content = getStatus('Source 1', GET_SOURCE1_URL, dateStr, '05:00')
lastS1Minutes = computeMinutes(dateStr, source1Content)
source2Content = getStatus('Source 2', GET_SOURCE2_URL, dateStr, '05:00')
lastS2Minutes = computeMinutes(dateStr, source2Content)

source1Dict, lastWrittenSource1Wh = \
    readInMemory('Source 1', source1Content, dateStr, lastTMinutes, 0)
source2Dict, lastWrittenSource2Wh = \
    readInMemory('Source 2', source2Content, dateStr, lastTMinutes, 0)

while True:
    log('Sleeping for 5 minutes')
    time.sleep(300)  # wait 5 minutes before checking again

    now = datetime.now()
    if now.hour < 5 or now.hour > 22:  # only check between 5 and 23 hours
        log('Outside solar generation hours (5..23)')
        sys.exit('Exiting program to avoid possible memory leaks')

    timeStr = minutesToTimeString(lastTMinutes)
    source1Content = getStatus('Source1', GET_SOURCE1_URL, dateStr, timeStr)
    lastS1Minutes = computeMinutes(dateStr, source1Content)
    source1Dict = {}
    if lastS1Minutes > lastTMinutes:
        source1Dict, nop1Wh = \
            readInMemory(
                'Source 1', source1Content, dateStr,
                lastTMinutes, lastWrittenSource1Wh
            )
    elif now.hour != 22:
        # when after 22:00 if there are results, process them anyway
        # because other inverter might be sleeping
        continue  # nothing to do

    source2Content = getStatus('Source2', GET_SOURCE2_URL, dateStr, timeStr)
    lastS2Minutes = computeMinutes(dateStr, source2Content)
    source2Dict = {}
    if lastS2Minutes > lastTMinutes:
        source2Dict, nop2Wh = \
            readInMemory(
                'Source 2', source2Content, dateStr,
                lastTMinutes, lastWrittenSource2Wh
            )
    elif now.hour != 22:
        # when after 22:00 if there are results, process them anyway
        # because other inverter might be sleeping
        continue  # nothing to do

    if lastS1Minutes <= lastTMinutes and lastS2Minutes <= lastTMinutes:
        continue  # nothing to do

    log("Processing....")
    dataPointsCount = 0
    dataPointsStr = 'data='
    source1Keys = list(source1Dict.keys())
    source2Keys = list(source2Dict.keys())
    # Since python 3.7 the order is insertion
    # so the minimum of first entries and last entries can be computed

    len1 = len(source1Keys)
    len2 = len(source2Keys)
    if len1 > 0 and len2 > 0:
        fromMinutes = min(source1Keys[0], source2Keys[0])
        toMinutes = min(source1Keys[len1-1], source2Keys[len2-1])
    elif len1 > 0:
        fromMinutes = source1Keys[0]
        toMinutes = source1Keys[len1-1]
    else:
        fromMinutes = source2Keys[0]
        toMinutes = source2Keys[len2-1]

    while fromMinutes <= toMinutes:
        d = dateStr
        t = minutesToTimeString(fromMinutes)
        wh = 0
        w = 0
        v = 0.0
        if fromMinutes in source1Dict:
            line1 = source1Dict[fromMinutes]
            # print('line1=[' + line1 + ']')
            d1, t1, whStr1, wStr1, nop1, nop2, temp1, vStr1 = line1.split(',')
            lastWrittenSource1Wh = int(whStr1)
            wh = lastWrittenSource1Wh
            w = int(wStr1)
            v = float(vStr1)
        else:
            wh = lastWrittenSource1Wh

        if fromMinutes in source2Dict:
            line2 = source2Dict[fromMinutes]
            # print('line2=[' + line2 + ']')
            d2, t2, whStr2, wStr2, nop3, nop4, temp2, vStr2 = line2.split(',')
            lastWrittenSource2Wh = int(whStr2)
            wh += lastWrittenSource2Wh
            w += int(wStr2)
            v += float(vStr2)
        else:
            wh = lastWrittenSource2Wh

        if fromMinutes > lastTMinutes:
            # only write when AFTER last target time
            pvoutputString = \
                d + ',' + t + ',' + \
                str(wh) + ',' + str(w) + ',,,,' + str(round(v))
            dataPointsCount += 1
            if dataPointsCount > 1:
                dataPointsStr += ';'
            dataPointsStr += pvoutputString
            if dataPointsCount >= 29:
                addBatchStatus(dataPointsCount, dataPointsStr)
                dataPointsStr = 'data='
                dataPointsCount = 0

        fromMinutes += 5  # next 5 minutes

    if dataPointsCount > 0:
        addBatchStatus(dataPointsCount, dataPointsStr)

    lastTMinutes = min(lastS1Minutes, lastS2Minutes)


combine_two_pvoutput_systems.py
# == combine_two_pvoutput_systems.py Author: Zuinige Rijder ==================
"""
Simple Python3 script to combine two PVOutput Systems continously
(normally once per 5 minutes) into target PVOutput System.
"""
import time
import sys
import configparser
import socket
import traceback
from datetime import datetime
from urllib.error import HTTPError, URLError
from urllib.request import urlopen, Request

# == read api_secrets in combine_two_pvoutput_systems.cfg ====================
parser = configparser.ConfigParser()
parser.read('combine_two_pvoutput_systems.cfg')
api_secrets = dict(parser.items('api_secrets'))

# == Secrets, fill in yours in combine_two_pvoutput_systems.cfg ==============
PVOUTPUT_SOURCE1_API_KEY = api_secrets['pvoutput_source1_api_key']
PVOUTPUT_SOURCE1_SYSTEM_ID = api_secrets['pvoutput_source1_system_id']

PVOUTPUT_SOURCE2_API_KEY = api_secrets['pvoutput_source2_api_key']
PVOUTPUT_SOURCE2_SYSTEM_ID = api_secrets['pvoutput_source2_system_id']

PVOUTPUT_TARGET_API_KEY = api_secrets['pvoutput_target_api_key']
PVOUTPUT_TARGET_SYSTEM_ID = api_secrets['pvoutput_target_system_id']

# == Constants ===============================================================
PVOUTPUT_GET_URL = 'https://pvoutput.org/service/r2/getstatus.jsp'
PVOUTPUT_ADD_URL = 'http://pvoutput.org/service/r2/addbatchstatus.jsp'

GET_SOURCE_1_URL = PVOUTPUT_GET_URL + \
    '?key=' + PVOUTPUT_SOURCE1_API_KEY + \
    '&sid=' + PVOUTPUT_SOURCE1_SYSTEM_ID + \
    '&h=1&asc=0&limit=288'
GET_SOURCE_2_URL = PVOUTPUT_GET_URL + \
    '?key=' + PVOUTPUT_SOURCE2_API_KEY + \
    '&sid=' + PVOUTPUT_SOURCE2_SYSTEM_ID + \
    '&h=1&asc=0&limit=288'
GET_TARGET_URL = PVOUTPUT_GET_URL + \
    '?key=' + PVOUTPUT_TARGET_API_KEY + \
    '&sid=' + PVOUTPUT_TARGET_SYSTEM_ID + \
    '&h=1&asc=0&limit=288'
TARGET_HEADERS = {
    'X-Pvoutput-Apikey': PVOUTPUT_TARGET_API_KEY,
    'X-Pvoutput-SystemId': PVOUTPUT_TARGET_SYSTEM_ID,
    'Content-Type': 'application/x-www-form-urlencoded',
    'Accept': 'text/plain'
}

TODAY = datetime.now().strftime("%Y%m%d")  # format yyyymmdd


# == log =====================================================================
def log(msg):
    """log a message prefixed with a date/time format yyyymmdd hh:mm:ss"""
    print(TODAY + datetime.now().strftime(" %H:%M:%S") + ': ' + msg)


# == print_splitted ==========================================================
def print_splitted(content):
    """print parameter splitted by ; line by line"""
    count = 0
    for item in content.split(';'):
        count += 1
        print(str(count) + ': ' + item)


# == execute_request =========================================================
def execute_request(url, request) -> str:
    """execute request and log url in case of errors"""
    errorstring = ''
    try:
        with urlopen(request, timeout=30) as response:
            body = response.read()
            content = body.decode("utf-8")
            return content
    except HTTPError as error:
        errorstring = str(error.status) + ': ' + error.reason
    except URLError as error:
        errorstring = str(error.reason)
    except TimeoutError:
        errorstring = 'Request timed out'
    except socket.timeout:
        errorstring = 'Socket timed out'
    except Exception as ex:  # pylint: disable=broad-except
        errorstring = 'urlopen exception: ' + str(ex)
        traceback.print_exc()

    log('ERROR: ' + url + ' -> ' + errorstring)
    time.sleep(60)  # retry after 1 minute

    return 'ERROR'


# == add_datapoints ==========================================================
def add_datapoints(datapoints_count, datapoints_str) -> str:
    """add datapoints"""
    log('Target datapoints: ' + str(datapoints_count))
    print_splitted(datapoints_str)
    post_data = datapoints_str.encode("utf-8")
    retry = 0
    while True:
        retry += 1
        request = Request(
            PVOUTPUT_ADD_URL, data=post_data, headers=TARGET_HEADERS)
        content = execute_request(PVOUTPUT_ADD_URL, request)
        if content != 'ERROR' or retry > 30:
            if content == 'ERROR':
                log('ERROR: number of retries exceeded')
            return content


# == get_pvoutput_entries_today ==============================================
def get_pvoutput_entries_today(url) -> str:
    """get pvoutput entries since time today"""
    request_url = url + '&d="' + TODAY + '"'
    while True:
        request = Request(request_url)
        content = execute_request(request_url, request)
        if content != 'ERROR':
            return content


# == compute_minutes =========================================================
def compute_minutes(content) -> int:
    """compute minutes for this day with the first time"""
    minutes = 300  # start at 5 hours
    if len(content) > 14 and content[0:8] == TODAY:
        minutes = round(int(content[9:11]) * 60 + int(content[12:14]))
    return minutes


# == minutes_to_hhmm =========================================================
def minutes_to_hhmm(minutes) -> str:
    """Convert minutes to timestring in format hh:mm"""
    hours = int(minutes / 60)
    rounded_minutes = round(minutes - (60 * hours))
    hhmm = f"{hours:02d}" + ':' + f"{rounded_minutes:02d}"
    return hhmm


# prev tuple index CONSTANTS
LINE_TUPLE_INDEX = 0
WH_TUPLE_INDEX = 1
VOLT_STR_TUPLE_INDEX = 2


# == process_one_pvoutput_result =============================================
def process_one_pvoutput_result(
    line,
    prev,  # tuple (prev_line, prev_wh, prev_volt_str)
    target_minutes,
    minutes_dict
) -> tuple[int, int, str, str]:
    """process one pvoutput result"""
    (date_str, time_str, watthour_str, _, watt_str,
        _, _, _, _, _, volt_str) = line.split(',')
    if date_str != TODAY:  # skip dates not today
        return 0, 0, '0', ''  # not interested in other days

    watthour = int(watthour_str)
    if volt_str == 'NaN':
        volt_str = prev[VOLT_STR_TUPLE_INDEX]  # just use previous volt value

    minutes = compute_minutes(line)
    if minutes % 5 != 0:
        sys.exit('Minutes not in 5 minutes increments: ' + line)

    line = date_str + ',' + \
        time_str + ',' + str(watthour) + ',' + watt_str + ',,,,' + volt_str
    if minutes > target_minutes:
        if prev[LINE_TUPLE_INDEX] == '':
            prev_minutes = minutes - 5  # no missing entries
        else:
            prev_minutes = compute_minutes(prev[LINE_TUPLE_INDEX])
            minutes_dict[prev_minutes] = prev[LINE_TUPLE_INDEX]

        if prev_minutes + 5 != minutes:
            print(
                'Missing entry:\n---> ' +
                prev[LINE_TUPLE_INDEX] + '\n---> ' + line)
            missing_entries = int(((minutes - prev_minutes - 5) / 5))
            watthour = int(
                prev[WH_TUPLE_INDEX] +
                ((watthour - prev[WH_TUPLE_INDEX]) / (1 + missing_entries))
            )
            time_str = minutes_to_hhmm(prev_minutes + 5)
            added_line = date_str + ',' + \
                time_str + ',' + str(watthour) + \
                ',' + watt_str + ',,,,' + volt_str

            print('Prev : ' + prev[LINE_TUPLE_INDEX])
            print('Added: ' + added_line)
            print('Next : ' + line + '\n')
            minutes_dict[prev_minutes + 5] = added_line
            line = added_line
        else:
            minutes_dict[minutes] = line

    return minutes, watthour, volt_str, line


# == read_entries_in_memory_with_5_minutes_intervals =========================
def read_entries_in_memory_with_5_minutes_intervals(
    prefix,
    content,
    written_target_minutes,
    written_watthour
) -> tuple[int, dict]:
    """read entries in memory with 5 minutes intervals"""
    since_time = minutes_to_hhmm(written_target_minutes)
    minutes_dict = {}
    splitted = content.split(';')
    prev_line = ''
    prev_wh = written_watthour
    prev_volt_str = '0'
    written_hhmm = since_time
    i = len(splitted) - 1
    while i >= 0:
        line = splitted[i]
        i -= 1

        prev_minutes = compute_minutes(prev_line)
        minutes, prev_wh, prev_volt_str, line = process_one_pvoutput_result(
            line, (prev_line, prev_wh, prev_volt_str), written_target_minutes,
            minutes_dict)
        if line == '':
            continue  # skip empty lines

        if minutes <= written_target_minutes:
            written_hhmm = minutes_to_hhmm(minutes)
            written_watthour = prev_wh
        else:
            if prev_line != '' and prev_minutes + 5 != minutes:
                i += 1  # do this entry again next loop

        prev_line = line

    log(prefix +
        ' read_in_memory since ' + since_time + ': written=' + written_hhmm +
        ', written wh=' + str(written_watthour) + ', wh=' + str(prev_wh) +
        ', line=' + prev_line)
    return written_watthour, minutes_dict


# == compute_values ==========================================================
def compute_values(
    source_dict, from_minutes, written_watthour
) -> tuple[int, int, float, int]:
    """compute watthour, watt, volt and last written watthour"""
    watthour = 0
    watt = 0
    volt = 0.0
    if from_minutes in source_dict:
        line = source_dict[from_minutes]
        _, _, watthour_str, watt_str, _, _, _, volt_str = line.split(',')
        written_watthour = int(watthour_str)
        watt = int(watt_str)
        volt = float(volt_str)

    watthour = written_watthour

    return watthour, watt, volt, written_watthour


# == compute_combined_values =================================================
def compute_combined_values(
    source_1_dict,
    source_2_dict,
    from_minutes,
    written_watthour_1,
    written_watthour_2
) -> tuple[int, int, float, int, int]:
    """compute combined values of source 1 and 2"""
    watthour, watt, volt, written_watthour_1 = compute_values(
        source_1_dict, from_minutes, written_watthour_1)

    watthour2, watt2, volt2, written_watthour_2 = compute_values(
        source_2_dict, from_minutes, written_watthour_2)

    watthour += watthour2
    watt += watt2
    volt += volt2

    return (
        watthour,
        watt,
        volt,
        written_watthour_1,
        written_watthour_2
    )


# == compute_from_and_to_minutes =============================================
def compute_from_and_to_minutes(
    source_1_dict, source_2_dict
) -> tuple[int, int]:
    """compute from_minutes and to_minutes"""

    # Since python 3.7 the order is insertion
    # so the minimum of first entries and last entries can be computed
    source_1_keys = list(source_1_dict.keys())
    source_2_keys = list(source_2_dict.keys())
    len1 = len(source_1_keys)
    len2 = len(source_2_keys)
    if len1 > 0 and len2 > 0:
        from_minutes = min(source_1_keys[0], source_2_keys[0])
        to_minutes = min(source_1_keys[len1-1], source_2_keys[len2-1])
    elif len1 > 0:
        from_minutes = source_1_keys[0]
        to_minutes = source_1_keys[len1-1]
    else:
        from_minutes = source_2_keys[0]
        to_minutes = source_2_keys[len2-1]

    return from_minutes, to_minutes


# == construct_and_add_datapoints ===========================================
def construct_and_add_datapoints(
    pvoutput_string, datapoints_count, datapoints_str
) -> tuple[int, str]:
    """construct and add datapoints if above 29"""
    datapoints_count += 1
    if datapoints_count > 1:
        datapoints_str += ';'
    datapoints_str += pvoutput_string
    if datapoints_count >= 29:
        add_datapoints(datapoints_count, datapoints_str)
        datapoints_str = 'data='
        datapoints_count = 0

    return datapoints_count, datapoints_str


# == process_sources_entries =================================================
def process_sources_entries(
    source_1_dict,
    source_2_dict,
    written_target_minutes,
    written_watthour_1,
    written_watthour_2
) -> tuple[int, int]:
    """process sources entries"""
    datapoints_count = 0
    datapoints_str = 'data='

    from_minutes, to_minutes = compute_from_and_to_minutes(
        source_1_dict, source_2_dict)
    while from_minutes <= to_minutes:
        watthour, watt, volt, written_watthour_1, written_watthour_2 = \
            compute_combined_values(
                source_1_dict,
                source_2_dict,
                from_minutes,
                written_watthour_1,
                written_watthour_2
            )

        # only write when AFTER last target time
        if from_minutes > written_target_minutes:
            pvoutput_string = (
                TODAY + ',' +
                minutes_to_hhmm(from_minutes) + ',' +
                str(watthour) + ',' + str(watt) + ',,,,' + str(round(volt))
            )
            datapoints_count, datapoints_str = construct_and_add_datapoints(
                pvoutput_string, datapoints_count, datapoints_str)

        from_minutes += 5  # next 5 minutes

    if datapoints_count > 0:
        add_datapoints(datapoints_count, datapoints_str)

    return written_watthour_1, written_watthour_2


# == compute_entries_since ===================================================
def compute_entries_since(
    prefix, url, written_target_minutes, written_watthour
) -> tuple[int, int, dict]:
    """compute dictionary entries since time"""
    content = get_pvoutput_entries_today(url)
    current_minutes = compute_minutes(content)
    minutes_dict = {}
    written_watthour = 0
    if current_minutes > written_target_minutes:
        written_watthour, minutes_dict = \
            read_entries_in_memory_with_5_minutes_intervals(
                prefix, content,
                written_target_minutes, written_watthour
            )

    return current_minutes, written_watthour, minutes_dict


# == main_loop ===============================================================
def main_loop():
    """main loop"""
    # get the latest values
    written_target_minutes, _, _ = compute_entries_since(
        'Target  ', GET_TARGET_URL, 300, 0)
    log("Last Written target: " + minutes_to_hhmm(written_target_minutes))
    written_minutes_1, written_watthour_1, _ = compute_entries_since(
        'Source 1', GET_SOURCE_1_URL, written_target_minutes, 0)
    written_minutes_2, written_watthour_2, _ = compute_entries_since(
        'Source 2', GET_SOURCE_2_URL, written_target_minutes, 0)
    log('Sleeping till next 5 minutes plus 3 minutes')

    while True:
        now = datetime.now()
        time.sleep(480 - (int(now.timestamp()) % 300))  # next 5+3 minutes incr

        # only check between 5 and 23 hours
        if now.hour < 5 or now.hour > 22:
            log('Outside solar generation hours (5..23)')
            sys.exit('Exiting program to start fresh tomorrow')

        written_minutes_1, _, source_1_dict = compute_entries_since(
            'Source 1', GET_SOURCE_1_URL, written_target_minutes,
            written_watthour_1)
        if written_minutes_1 <= written_target_minutes and now.hour != 22:
            # when after 22:00 if there are results, process them anyway
            # because other inverter might be sleeping
            continue  # nothing to do

        written_minutes_2, _, source_2_dict = compute_entries_since(
            'Source 2', GET_SOURCE_2_URL, written_target_minutes,
            written_watthour_2)
        if written_minutes_2 <= written_target_minutes and now.hour != 22:
            # when after 22:00 if there are results, process them anyway
            # because other inverter might be sleeping
            continue  # nothing to do

        if (written_minutes_1 <= written_target_minutes and
                written_minutes_2 <= written_target_minutes):
            continue  # nothing to do

        written_watthour_1, written_watthour_2 = \
            process_sources_entries(
                source_1_dict,
                source_2_dict,
                written_target_minutes,
                written_watthour_1,
                written_watthour_2
            )

        written_target_minutes = min(written_minutes_1, written_minutes_2)


# == MAIN ====================================================================
main_loop()
